### ISSUE TITLE
DrGRPO Implement is right???,  the loss is micro_batch_size * loss, and the grad value is micro_batch_size * grad ???  why ??? is it satisfies the statistical theorem algorithm

### ISSUE THREAD

**Issue by l1351868270** [verl-project/verl#3741]
### System Info

master

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

The now implement is
```
elif loss_agg_mode == "seq-mean-token-sum-norm":
        seq_losses = torch.sum(loss_mat * loss_mask, dim=-1)
        loss = torch.sum(seq_losses) / loss_mask.shape[-1]  # The divisor
        # (loss_mask.shape[-1]) should ideally be constant
        # throughout training to well-replicate the DrGRPO paper.
        # TODO: Perhaps add user-defined normalizer argument to
        # agg_loss to ensure divisor stays constant throughout.
```

### Expected behavior

why not
```
elif loss_agg_mode == "seq-mean-token-sum-norm":
        seq_losses = torch.sum(loss_mat * loss_mask, dim=-1)
        loss = torch.mean(seq_losses) / loss_mask.shape[-1]  # The divisor
        # (loss_mask.shape[-1]) should ideally be constant
        # throughout training to well-replicate the DrGRPO paper.
        # TODO: Perhaps add user-defined normalizer argument to
        # agg_loss to ensure divisor stays constant throughout.
```
or
```
elif loss_agg_mode == "seq-mean-token-sum-norm":
        seq_losses = torch.sum(loss_mat * loss_mask, dim=-1)  / loss_mask.shape[-1]  # The divisor
        loss = torch.mean(seq_losses)
        # (loss_mask.shape[-1]) should ideally be constant
        # throughout training to well-replicate the DrGRPO paper.
        # TODO: Perhaps add user-defined normalizer argument to
        # agg_loss to ensure divisor stays constant throughout.
```

**Comment by lwp-hhxx** [comment]
i have the same question

**Comment by haz269** [comment]
i have the same question

**Comment by Kirrito-k423** [comment]
Thanks for raising this. The code snippet in this issue appears to be from an older implementation. On current `main`, `seq-mean-token-sum-norm` no longer computes only `sum(seq_losses) / horizon`.

Current behavior in `verl/trainer/ppo/core_algos.py` is:

1. sum token losses within each response: `seq_losses = sum(loss_mat * loss_mask, dim=-1)`;
2. average those response-level sums over the global PPO mini-batch: `masked_sum(seq_losses, seq_mask) / global_batch_size * dp_size`;
3. divide by `loss_scale_factor`, or by `loss_mask.shape[-1]` when `loss_scale_factor` is not set.

So in the single-DP case this is equivalent to `torch.mean(seq_losses) / C`, where `C` is `loss_scale_factor` or the response horizon. It should not scale the loss/gradient by the local micro batch size in the way shown in the original report.

Relevant merged changes:
- #4360 added `loss_scale_factor` for this mode.
- #5366 made the aggregation parallelism-invariant when DP is used.
- #5417 fixed the missing seq-mean semantics for `seq-mean-token-sum-norm`.

中文说明：这个 issue 里引用的片段是旧实现。当前 `main` 中 `seq-mean-token-sum-norm` 会先对每个 response 做 token-sum，再按全局 PPO mini-batch 的 sequence 数做平均，最后除以 `loss_scale_factor`（未设置时使用 response horizon）。因此单 DP 情况下等价于 `torch.mean(seq_losses) / C`，不会像原问题里担心的那样按本地 micro batch size 放大 loss 或梯度。

This response was generated by a scheduled Codex GPT5.5 issue processing task.
